packages/asset-swapper/contracts/src/NativeOrderSampler.sol

```diff
@@ -0,0 +1,202 @@
+        IERC20Token makerToken = _getTokenFromERC20AssetData(order.makerAssetData);
+        IERC20Token makerFeeToken = order.makerFee > 0
+            ? _getTokenFromERC20AssetData(order.makerFeeAssetData)
+            : IERC20Token(0);
+        uint256 remainingTakerAmount = order.takerAssetAmount
+            .safeSub(orderInfo.orderTakerAssetFilledAmount);
+        fillableTakerAmount = remainingTakerAmount;
+        // The total fillable maker amount is the remaining fillable maker amount
+        // PLUS maker fees, if maker fees are denominated in the maker token.
+        uint256 totalFillableMakerAmount = LibMath.safeGetPartialAmountFloor(
+            remainingTakerAmount,
+            order.takerAssetAmount,
+            makerFeeToken == makerToken
+                ? order.makerAssetAmount.safeAdd(order.makerFee)
+                : order.makerAssetAmount
+        );
+        // The spendable amount of maker tokens (by the maker) is the lesser of
+        // the maker's balance and the allowance they've granted to the ERC20Proxy.
+        uint256 spendableMakerAmount = LibSafeMath.min256(
+            makerToken.balanceOf(order.makerAddress),
+            makerToken.allowance(order.makerAddress, spender)
+        );
+        // Scale the fillable taker amount by the ratio of the maker's
+        // spendable maker amount over the total fillable maker amount.
+        if (spendableMakerAmount < totalFillableMakerAmount) {
+            fillableTakerAmount = LibMath.getPartialAmountCeil(
+                spendableMakerAmount,
+                totalFillableMakerAmount,
+                remainingTakerAmount
+            );
+        }
+        // If the maker fee is denominated in another token, constrain
+        // the fillable taker amount by how much the maker can pay of that token.
+        if (makerFeeToken != makerToken && makerFeeToken != IERC20Token(0)) {
+            uint256 spendableExtraMakerFeeAmount = LibSafeMath.min256(
+                makerFeeToken.balanceOf(order.makerAddress),
+                makerFeeToken.allowance(order.makerAddress, spender)
+            );
+            if (spendableExtraMakerFeeAmount < order.makerFee) {
+                fillableTakerAmount = LibSafeMath.min256(
+                    fillableTakerAmount,
+                    LibMath.getPartialAmountCeil(
+                        spendableExtraMakerFeeAmount,
+                        order.makerFee,
+                        remainingTakerAmount
+                    )
+                );
+            }
+        }
+    }
+
+    function _getTokenFromERC20AssetData(bytes memory assetData)
+        private
+        pure
+        returns (IERC20Token token)
+    {
+        if (assetData.length == 0) {
+            return IERC20Token(address(0));
+        }
+        if (assetData.length != 36 ||
+            assetData.readBytes4(0) != ERC20_ASSET_PROXY_ID)
+        {
+            revert('NativeOrderSampler/UNSUPPORTED_ASSET_DATA');
```

Comment:
can we do `address(0)` and return `0` if it's unsupported? Always concerned about a bad order getting in and messing the whole sampler result.